db/view/view_update_generator.cc

```diff
@@ -437,6 +437,15 @@ future<> view_update_generator::generate_and_propagate_view_updates(const replic
 
             co_await seastar::sleep(throttle_delay);
 
+            if (_db.view_update_sem().current() == 0) {
+                utils::get_local_injector().inject("view_update_limit", [&err] {
+                    err = std::make_exception_ptr(exceptions::view_update_generation_timeout_exception());
+                });
+                if (err) {
+                    break;
+                }
+            }
```

Comment:
When you add error injection code, it should be able for the compiler to optimize it out if injections are disabled by the build mode. Now, it's not the case because you are adding two nested `if`s that will execute even if the injection is not enabled.

I suppose you wrote it like that because you didn't know how to handle flow control (inject is a lambda so you can't really break the outer loop from it)? Consider using `utils::error_injection::enter` function to realize your goal instead.

Reply:
I must admit that I don't understand the purpose of the error injection here. This will trigger a timeout error only after processing the second batch and when there are _no_ view updates in progress.

Could you explain what it is for and include this explanation in the comment? It would also be great if the test scenario referred to it.

Reply:
> When you add error injection code, it should be able for the compiler to optimize it out if injections are disabled by the build mode. Now, it's not the case because you are adding two nested `if`s that will execute even if the injection is not enabled. The code will actually execute every time.
> 
> I suppose you wrote it like that because you didn't know how to handle flow control (inject is a lambda so you can't really break the outer loop from it)? Consider using `utils::error_injection::enter` function to realize your goal instead.

I didn't really like the way I implemented this either, thanks for pointing to `utils::error_injection::enter`, it's just what was needed

> I must admit that I don't understand the purpose of the error injection here. This will trigger a timeout error only after processing the second batch and when there are _no_ view updates in progress.

I added the injection only for the second (and following) batch because we're already failing requests if we encounter a full backlog due to #17777

The error will be thrown when there are too many view updates in progress, i.e. when the free view_update_semaphore units will be 0 (this can also be check using the view backlog (and check if `relative_size()` is 0), but IIRC it doesn't have public methods so it would require more changes.

> Could you explain what it is for and include this explanation in the comment? It would also be great if the test scenario referred to it.

Ok